### ISSUE TITLE
[meta] Big questions

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#48]
List of big questions to resolve for the content of the API (not the shape):
  1. **resource allocation**
      - how are resources bound to memory
      - whether memory heaps are exposed or implied
  2. **resource capabilities** - #37
      - do we ask the user for the set of potential usages for each resource (akin [D3D12_RESOURCE_FLAGS](https://msdn.microsoft.com/en-us/library/windows/desktop/dn986742(v=vs.85).aspx)) or imply this
  3. **multiple queues**
      - do we expose multiple hardware queues (limited in Vulkan, unlimited in D3D12) or try to parallelize in the driver itself (like Metal)
  4. **memory barriers**- #27:
      - choices: implicit, half-explicit (with only the destination state provided), fully explicit (Vulkan, D3D12)
  5. **partial pipeline states**
      - do we create blend, depth-stencil, raterizer (and other) states independently, or only as a part of a bigger pipeline state?
      - support for non-dynamic states (Vulkan only)
  6. **tile memory**
      - do we take advantage of tiled renderers?
      - existing APIs: tile memory in Metal, multiple sub-passes in Vulkan
      - blocked by D3D12
  7. **data upload/download** - #45
      - persistent mapping support
      - coherency
      - callbacks versus polling
  8. **web workers support**
      - resource creation
      - command buffer recording
  9. **shader languages** - #42,#43,#44:
      - consume SPIR-V and/or HLSL
      - support for pipeline/shader caches
  10. **canvas interop**
      - swapchain initialization, frame sync
      - webXR integration
  11. **security == portability** (~~N==NP~~):
      - is there a portability trade-off with performance that we can take without compromising the security?

**Comment by kainino0x** [comment]
Editing to also note other DOM integration.

**Comment by kvark** [comment]
Got some detailed feedback from @zeux on some of the big questions. Arseny Kapoulkine has a great deal of experience in low-level graphics optimization of Vulkan and Metal-1 on both mobile and desktop.
> At Roblox we have a pretty carefully optimized renderer that uses Metal & Vulkan; we don't necessarily strongly favor one or the other.

####  1. resource allocation
> allocation performance has been very much an issue for us in both Metal and DX11, and has been a non-issue in Vulkan. In Vulkan we know allocations are slow, but they also can be large and we can easily suballocate and manage our own memory - crucially, without conflating the lifetime of individual objects sharing the space or having to worry much about compatibility requirements
> FWIW we don't have separate codepaths for Vulkan for NV/etc. - we just have generic code that looks at available memory types and picks the one that the driver exposes that matches our requirements for the resource. We do have special cases for devices that expose device-local&host-available mode (mobile/unified memory), but otherwise we just deal with whatever the driver throws at us.

Thus, we need some sort of heap allocation for resources. Whether it's restricted as Metal or free-style as Vulkan is up for discussion.

#### 3. multiple queues
> as far as I understand is impossible in Metal to this day is compute overlap (e.g. having several compute dispatches read & write the same resource concurrently, provided that the application guarantees lack of race conditions) - we had some prototypes for algorithms where this was significant for GPU performance (double-digit percentages), but that should be possible to achieve in Metal with small API tweaks, similar to how NV/AMD have extensions for this in DX11

Thus, we either need to figure out what those "small API tweaks" are, or expose multiple queues like D3D12/Vulkan. A follow up question to Apple - any idea on why Metal driver can't do async compute?

#### 4. memory barriers
> While explicit barriers in Vulkan allow the app to implement any management strategy it desires to cut down on the CPU/GPU cost, they are way more complicated than what would be ideal for an explicit design. They are hard to understand, hard to reason about, lots of people get them wrong - including in vendor SDK samples, sometimes vendor articles, etc. - *and* validation layers don't validate their correctness at all at this time. Whether explicit or implicit barriers are better is debatable, for explicit barriers I would definitely recommend the resource state model, as used by DX12 and for example this library https://github.com/Tobski/simple_vulkan_synchronization (I believe the author had a significant part in the barrier design in Vulkan, the fact that he made this library might be telling :D). DX12 does get some things wrong, such as lack of UNDEFINED state transitions and potentially insufficient state coverage as applicable to tilers, but the model itself is much simpler. This is a big factor in both reducing application complexity, and reducing potential for driver bugs - we definitely have seen Vulkan bugs where drivers misinterpreted the (valid) barrier setup as well.

TL;DR: we don't have any performance data on the effectiveness of explicit barriers, but we know that they are super hard to get right.

#### 9. shader languages
> our shaders are written in HLSL; our shader translation pipeline is slowly moving towards being entirely SPIRV based for GL/Metal/Vulkan

This info doesn't show us a way forward, but at least confirms that we are on the right track :)

#### (unlisted) resource binding - #19
> For draw call cost in general, Vulkan binding model *should* have an advantage compared to Metal for some architectures in terms of CPU update cost - Metal uses per-stage indexed bindings, which doesn't really match AMD designs for one, and requires driver to shadow a lot of that state and build actual HW descriptors at draw call time. There are a lot of other design decisions in Vulkan that allow to reduce binding cost further in the driver, for example buffer object descriptors can - on some hardware - be packed into the 64-bit object handle which means that when you update the descriptors, the driver can copy handles from your memory into, say, GPU memory or registers, without any cache misses that are required in Metal model - unless you group all your buffer objects into very few very big ones. Again, future Metal revisions try to improve on this significantly, specifically with Metal 2's argument buffers, but we haven't used them so can't comment on viability

This confirms our choice of binding model to be a good. Experiments in gfx-rs have also confirmed that Metal 2's argument buffers can serve as an efficient backend implementation, when available.

---
Ok, I think this captures the essence of Arseny's feedback (thank you!). Please feel free to correct me or add more info. I tried to stay objective and communicate information that is most helpful for the group to move forward, not necessarily aligning with Mozilla's prototype. The conclusions I drew are up for debate, and we'll surely need detailed feedback from other ISVs in order to make a good judgement.

**Comment by Kangz** [comment]
This is great feedback, and I have a bunch of questions / comments on it, however I'd rather do it in another issue / PR. Do you think you can move it?

**Comment by kvark** [comment]
Yep. We haven't been updating this anyway. It's good as a historical reference.
